File: src/0673_number_of_longest_increasing_subsequence.js
function findNumberOfLIS(nums) {
    const dp = new Map();
    let lenLIS = 0;
    let res = 0;
    const dfs = (i) => {
        if (dp.has(i)) return dp.get(i);
        let maxLen = 1;
        let maxCnt = 1;
        for (let j = i + 1; j < nums.length; j++) {
            if (nums[j] > nums[i]) {
                const [length, count] = dfs(j);
                if (length + 1 > maxLen) {
                    maxLen = length + 1;
                    maxCnt = count;
                } else if (length + 1 === maxLen) {
                    maxCnt += count;
                }
            }
        }
        if (maxLen > lenLIS) {
            lenLIS = maxLen;
            res = maxCnt;
        } else if (maxLen === lenLIS) {
            res += maxCnt;
        }
        const value = [maxLen, maxCnt];
        dp.set(i, value);
        return value;
    };
    for (let i = 0; i < nums.length; i++) dfs(i);
    return res;
}

module.exports = { findNumberOfLIS };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(findNumberOfLIS([1, 3, 5, 4, 7]), 2);
    assert.deepStrictEqual(findNumberOfLIS([2, 2, 2, 2, 2]), 5);
}

File: src/0673_number_of_longest_increasing_subsequence.rs
use std::collections::HashMap;

fn lis_dfs(
    i: usize,
    nums: &[i32],
    dp: &mut HashMap<usize, (i32, i32)>,
    len_lis: &mut i32,
    res: &mut i32,
) -> (i32, i32) {
    if let Some(&cached) = dp.get(&i) {
        return cached;
    }

    let mut max_len = 1;
    let mut max_cnt = 1;
    for j in i + 1..nums.len() {
        if nums[j] > nums[i] {
            let (length, count) = lis_dfs(j, nums, dp, len_lis, res);
            if length + 1 > max_len {
                max_len = length + 1;
                max_cnt = count;
            } else if length + 1 == max_len {
                max_cnt += count;
            }
        }
    }

    if max_len > *len_lis {
        *len_lis = max_len;
        *res = max_cnt;
    } else if max_len == *len_lis {
        *res += max_cnt;
    }

    dp.insert(i, (max_len, max_cnt));
    (max_len, max_cnt)
}

pub fn find_number_of_lis(nums: Vec<i32>) -> i32 {
    let mut dp = HashMap::new();
    let mut len_lis = 0;
    let mut res = 0;

    for i in 0..nums.len() {
        lis_dfs(i, &nums, &mut dp, &mut len_lis, &mut res);
    }

    res
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mirrors_python_main_assertions() {
        assert_eq!(find_number_of_lis(vec![1, 3, 5, 4, 7]), 2);
        assert_eq!(find_number_of_lis(vec![2, 2, 2, 2, 2]), 5);
    }
}
